### ISSUE TITLE
sge qmod will kill nextflow running jobs

### ISSUE THREAD

**Issue by Crabime** [nextflow-io/nextflow#1001]
## Bug report 

(Job on SGE will be killed but not suspend if SGE already reached its threshold and submitted through Nextflow. Below  i made an demo and demonstrate such condition)

### Expected behavior and actual behavior

(hope running job can be suspend but not kill directly. it actual return 138 error and nextflow main process exit with error)

### Steps to reproduce the problem

(

1.  `qconf -sq all.q`, change `suspend_thresholds `value to `np_load_avg=0.6` and `nsuspend ` value 1
2.  run below time-consuming.nf script using command: `nextflow run time-consuming.nf --s 4`. attention: --s 4 because of my pc cpu num is 4 and each process consume 100% cpu, so if your server different with me, you should give `s` value more than cpu num so that `np_load_avg` will larger than 1 and SGE will reach threshold.
3.  wait for about a minute and nextflow will error and exit.
4. such process can be simply using another way. running nextflow the same way above and run command `qmod -sj yourSubmitTaskId`, then above error will occur
5. In order to prove that such error just happened if job submitted through nextflow, i convert such nextflow script to a simple bash script and qsub directly. run command `test-comsuming.sh 4`, after job submitted to SGE successfully and  `qmod -sj jobId` and `qstat`, you can see such job status became `s` but not exit.
)

### Program output 

(N E X T F L O W  ~  version 18.11.0-edge
Launching `/home/crabime/ideaProjects/nextflow-gooalgene/tests/time-consuming.nf` [sick_bose] - revision: 8e9e8906c4
[warm up] executor > sge
[4a/2089e2] Submitted process > testTimeConsuming (1)
[9b/7bb470] Submitted process > testTimeConsuming (2)
Error executing process > 'testTimeConsuming (1)'

Caused by:
  Process `testTimeConsuming (1)` terminated with an error exit status (138)

Command executed:

  stress -c 1 -t 600

Command exit status:
  138

Command output:
  stress: info: [11940] dispatching hogs: 1 cpu, 0 io, 0 vm, 0 hdd

Command wrapper:
  stress: info: [11940] dispatching hogs: 1 cpu, 0 io, 0 vm, 0 hdd

Work dir:
  /home/crabime/ideaProjects/nextflow-gooalgene/work/4a/2089e224ee3ba9d0dfb76af330122d

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`

WARN: Killing pending tasks (1))

### Environment 

* Nextflow version: [all] 
* Java version: [openjdk 1.8.0_191]
* Operating system: [centos 7]

### Additional context

(all file already zipped and upload)
[nextflowbug.zip](https://github.com/nextflow-io/nextflow/files/2767611/nextflowbug.zip)

**Comment by pditommaso** [comment]
My understanding is that SGE kill the job and it returns 138. Is that right? Being so NF will manage it as a failed task.

**Comment by Crabime** [comment]
No, 138 = 128 + 10, and 10 is the value of signal USR1. I glanced SGE source code and its doc and found below:
there is the sentence of sge [doc](http://gridscheduler.sourceforge.net/htmlman/htmlman5/queue_conf.html?pathrev=V62u5_TAG) about queue :

> suspend_method
  resume_method
  terminate_method
     These parameters can be used  for  overwriting  the  default
     method  used by Sun Grid Engine for suspension, release of a
     suspension and for termination of a job.  Per  default,  the
     signals  SIGSTOP,  SIGCONT  and SIGKILL are delivered to the
     job to perform these actions. However, for some applications
     this is not appropriate.

so if using qmod or SGE default behavior suspending jobs while exceed threshold, signal *SIGSTOP* will be send, but *SIGSTOP* value is 20 so it do nothing with signal USR1, but while i checked SGE source code and find it's signal handling function below:
![sigusr1](https://user-images.githubusercontent.com/18657762/51320583-40bea600-1a9b-11e9-86a6-4cd9989b5056.png)
So USR1 signal is real matter.
Next, i dig into nextflow source code and found `BashWrapperBuilder` code snippet below:
```
on_term() {
     set +e
    __KILL_CMD__
}
trap on_exit EXIT
trap on_term TERM INT USR1 USR2
```
I don't know why nextflow trap two user defined signal USR1 and USR2 and kill such process which will lead Nextflow main process exit ?

**Comment by pditommaso** [comment]
Because SIGUSR1 is also used to signal the job exceeded the soft resource request, therefore NF kill it. 

```
RESOURCE LIMITS
     The first two resource limit parameters, s_rt and h_rt,  are
     implemented  by Sun Grid Engine. They define the "real time"
     or also called "elapsed" or "wall clock" time having  passed
     since  the  start  of  the job. If h_rt is exceeded by a job
     running in the queue, it is aborted via the  SIGKILL  signal
     (see  kill(1)).   If  s_rt  is  exceeded,  the  job is first
     "warned" via the SIGUSR1 signal (which can be caught by  the
     job) and finally aborted after the notification time defined
     in the queue configuration parameter notify
```

http://gridscheduler.sourceforge.net/htmlman/htmlman5/queue_conf.html

Even though I could agree that a bit too hard.

**Comment by Crabime** [comment]
eh...there might some trouble on controlling cluster host load. If all jobs submitted through nextflow, after some more time, multiple jobs result to exceed SGE threshold and SGE will suspend such running job, what we need is just **suspend job**, in other word, this job status should be `s`, but actually nextflow will exit totally, so task failed, but it sounds not properly. 
Here i have no way but cancel threshold option and so that  cluster server status can not  guarantee, you know multiple jobs will let server `np_load_avg` to a large number, then server will be blocked, some blocked  job will always running on SGE without an end. I tried `strace` such block process and got all timeout. So i think cluster threshold should be set, but it conflict with our NF, can you give me some hint? thanks:pray:

**Comment by pditommaso** [comment]
Yes, I think you are right, the termination on SIGUSR1 can be removed.
